### ISSUE TITLE
`get_data()` omits variables created in a non-global environment

### ISSUE THREAD

**Issue by capnrefsmmat** [easystats/insight#760]
I'm not sure if the title is completely correct, but it's what I've deduced so far. Here's a test script:

```r
library(insight)

foo <- function() {
  mtcars$cylinders <- factor(mtcars$cyl)

  fit <- lm(mpg ~ cylinders * disp + hp, data = mtcars)

  return(get_data(fit))
}

cat("Fit inside other environment:\n")
print(names(foo()))

cat("In global environment:\n")

mtcars$cylinders <- factor(mtcars$cyl)

fit <- lm(mpg ~ cylinders * disp + hp, data = mtcars)

print(names(get_data(fit)))
```

This yields the following output:

```
Fit inside other environment:
[1] "mpg"  "disp" "hp"  
In global environment:
[1] "mpg"       "cylinders" "disp"      "hp
```

Notice that `cylinders` is present in both models, but only appears in the `get_data()` output when it's created in the global environment. And yet `get_data()` is called from the same environment as the model is fit in both cases, so it's not like we're doing wacky environment hijinks.

I think the cause may be related to #666, because this problem appeared when I upgraded from 0.18.8 to 0.19.1, and that seems like a relevant change during that time period. I happened to run into the problem in a testthat test case, but as the example above shows, it need not be in a test case to occur.

I haven't tried to narrow this down further, but I can test out specific package versions or changes if that'd help.

**Comment by vincentarelbundock** [comment]
@strengejacke will have the definitive word on this, but I can still fill in a few missing links.

TLDR: try the `source="frame"` argument. It may (or may not) work in simple models.

Longer:

In older versions `insight::get_data` used to extract the model frame, check each column names to see if they had been transformed, and then try to figure out the appropriate inverse transform to return to the original state.

The benefit of this is that we could extract data created in weird nested environments. One downside is that this was hard to maintain because there are many possible transformations and *many* corner cases to handle. One even more important downside was that if the corner cases were not properly handled, `get_data()` returned *incorrect* numerical results. This is obviously worst than not returning any data at all...

So in recent version, a decision was made to try to find the original data by iterating through environments until Global. When there's no record of the environment in which the model was fit, `insight` doesn't know where to look, and falls back on the Global version, which is probably what you are getting.

In the docs, I see an argument `source="frame"`, but I have not tried it myself. If you do, please be aware that this is *less safe*, and that it may not work as expected for complex models (especially those with transformations).

The other hack, of course, is to assign your new variable to a parent frame of your function with double arrow assignement: `<<-` instead of single arrow `<-`.

**Comment by capnrefsmmat** [comment]
Yes, adding `source = "frame"` solves the immediate problem, which is good. But not the larger problem: I'm actually trying to use `get_predictors()`, which calls `get_data()`, and `get_predictors()` doesn't let me specify the `source` argument for its call to `get_data()`.

Looking further at the code, I see two potential changes that would fix this issue. Here's a diff for `.recover_data_from_environment`:

```diff
diff --git a/R/get_data.R b/R/get_data.R
index 8ed0b8726..8732a0a41 100644
--- a/R/get_data.R
+++ b/R/get_data.R
@@ -210,11 +210,13 @@ get_data <- function(x, ...) {
   }
 
   # first, try environment of formula, see #666
-  dat <- .safe(eval(model_call$data, envir = environment(model_call$formula)))
+  dat <- .safe(eval(model_call$data, envir = environment(model_call$formula),
+                    enclos = NULL))
 
   # second, try to extract formula directly
   if (is.null(dat)) {
-    dat <- .safe(eval(model_call$data, envir = environment(find_formula(x))))
+    dat <- .safe(eval(model_call$data, envir = environment(find_formula(x)$conditional),
+                      enclos = NULL))
   }
 
   # next try, parent frame
```

There are two issues:

1. In the first call to `eval()`, `environment(model_call$formula)` is `NULL`. Consequently, R defaults to looking into the enclosure, which defaults to `parent.frame()`. So it looks in the parent frame and finds `mtcars`, because `mtcars` is defined in base R. It's just not the right one. We want really want it to fail (because the environment is null) and proceed to checking the formula, so we can tell it to use no enclosing environment.
2. `environment(find_formula(x))` should never work, right? Because `find_formula(x)` returns a list, not a function or formula, and so it does not have anything for `environment()` to return. Here I've used `$conditional` to get the actual formula (which *does* have the correct environment), but another way to do it would be `environment(formula(x))`. I just don't know if that'd be compatible with everything insight tries to be compatible with.

Perhaps one useful starting point would be to cook up some unit tests for various possibilities? Then it'd be easier to tell if these changes are correct. I could try making a few for `test-get_data.R` if that would help.

**Comment by vincentarelbundock** [comment]
Yes, tests would certainly be useful.

Unfortunately, I'm travelling now and won't be able to look at this for at least a couple weeks, but perhaps one of the active maintainers will pitch in sooner.

**Comment by strengejacke** [comment]
> I'm actually trying to use get_predictors(), which calls get_data(), and get_predictors() doesn't let me specify the source argument for its call to get_data().

If `get_data()` fails to find something in the environment, it falls back to looking for the model frame. But yes, first attempt is always the environment.

Your changes look good to me! Feel free to open an PR. Tests would be great, too, I'm just not sure how to test these different environment issues properly?

**Comment by vincentarelbundock** [comment]
Thanks for this! It's been a pain point for me, and I'm glad to see this part of the code base improved!

**Comment by strengejacke** [comment]
Yes, thanks for this great PR! :-)
